### PR TITLE
Auth handler - Support cloud.redhat.com authentication

### PR DESCRIPTION
**Pull request by nmagnezi**
This PR was ported from [bm-inventory PR-363](https://github.com/filanov/bm-inventory/pull/363)

### CODE REVIEW

**Comment by nmagnezi** [comment]
@ronniel1 New PR is here.

File: subsystem/utils_test.go [hunk 1]
```diff
@@ -27,7 +27,7 @@ func strToUUID(s string) *strfmt.UUID {
 }
 
 func registerHost(clusterID strfmt.UUID) *models.Host {
-	host, err := bmclient.Installer.RegisterHost(context.Background(), &installer.RegisterHostParams{
+	host, err := agentBMClient.Installer.RegisterHost(context.Background(), &installer.RegisterHostParams{
```
Comment (eranco74) [inline subsystem/utils_test.go:30]: Need to add a few more calls to this test just to make sure agent can call these APIs
GET  /clusters/{cluster_id}/downloads/files
POST /clusters/{cluster_id}/uploads/ingress-cert

File: pkg/auth/auth_handler_test.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+			expectedTokenValue:     userKeyValue,
+			authInfo:               UserAuthHeaderWriter(userKeyValue),
+			isListOperation:        true,
+			enableAuth:             true,
+			addHeaders:             true,
+			expectedRequestSuccess: true,
+		},
+		{
+			name:                   "Fail auth without headers",
+			expectedTokenValue:     agentKeyValue,
```
Comment (ronniel1) [inline pkg/auth/auth_handler_test.go:84]: Need to test successful user and agent auth
Also unsuccessful user auth (for now all agent auth should be succesfull, until @rollandf adds his code)
also need to test case where there are no headers supplied for either agent or user auth

File: Makefile [hunk 3]
```diff
@@ -118,7 +118,7 @@ deploy-inventory-service-file: deploy-namespace
 	sleep 5;  # wait for service to get an address
 
 deploy-service-requirements: deploy-namespace deploy-inventory-service-file
-	python3 ./tools/deploy_assisted_installer_configmap.py --target "$(TARGET)" --domain "$(INGRESS_DOMAIN)" --base-dns-domains "$(BASE_DNS_DOMAINS)" --namespace "$(NAMESPACE)" $(DEPLOY_TAG_OPTION)
+	python3 ./tools/deploy_assisted_installer_configmap.py --target "$(TARGET)" --domain "$(INGRESS_DOMAIN)" --base-dns-domains "$(BASE_DNS_DOMAINS)" --namespace "$(NAMESPACE)" $(DEPLOY_TAG_OPTION) --enable-auth "$(ENABLE_AUTH)"
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply Makefile:121]: Done in [PR-53](https://github.com/openshift/assisted-service/pull/53/files#diff-b67911656ef5d18c4ae36cb6741b7965R24)

File: Makefile [hunk 4]
```diff
@@ -131,19 +131,24 @@ deploy-postgres: deploy-namespace
...
+	export SERVICE=minikube-local-registry/assisted-service:minikube-test && export TEST_FLAGS=--subsystem-test && export ENABLE_AUTH="True" \
+	&& $(MAKE) update-minikube deploy-all
 
 ########
 # Test #
 ########
 
 subsystem-run: test subsystem-clean
 
+generate-keys:
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply Makefile:143]: Added it as a dependency for `deploy-test` so we can get the local test cert to the server when it starts.
The relevant call chain is as follows:

`deploy-test (now calling generate-keys) --> deploy-all --> deploy-service --> deploy-service-requirements`

the last one invokes a script named `tools/deploy_assisted_installer.py` which will [read the test certificate if present](https://github.com/openshift/assisted-service/blob/master/tools/deploy_assisted_installer.py#L44) (called only in subsystem-tests)

The changes are in [PR-70](https://github.com/openshift/assisted-service/pull/70) Makefile

File: pkg/auth/auth_handler_test.go (hunk 2 above)
Comment (nmagnezi) [reply pkg/auth/auth_handler_test.go:84]: Done.
As requested, added more unit tests in [PR-70](https://github.com/openshift/assisted-service/pull/70/files) `pkg/auth/auth_handler_test.go `

File: subsystem/utils_test.go (hunk 1 above)
Comment (nmagnezi) [reply subsystem/utils_test.go:30]: Done.
1. GET `DownloadClusterFiles` will now get called with **agent auth** in subsystem test `[only_k8s]install download_config_files`
2. POST `UploadClusterIngressCert` using **agent auth** is already being used in`cluster_test.go`([here for example](https://github.com/openshift/assisted-service/blob/master/subsystem/cluster_test.go#L732-L742))

Comment (nmagnezi) [reply subsystem/utils_test.go:30]: Implemented in [PR-70](https://github.com/openshift/assisted-service/pull/70/files) `subsystem/cluster_test.go `

**Pull request merged**